siriuspy/siriuspy/devices/idff.py

```diff
@@ -0,0 +1,302 @@
+
+from ..namesys import SiriusPVName as _SiriusPVName
+from ..search import IDSearch as _IDSearch
+from ..magnet.idffwd import APUFFWDCalc as _APUFFWDCalc
+from ..idff.config import IDFFConfig as _IDFFConfig
+
+from .device import Device as _Device, Devices as _Devices
+from .pwrsupply import PowerSupply as _PowerSupply
+from .ids import WIG as _WIG, APU as _APU
+
+
+class IDFF(_Devices):
+    """Insertion Device Feedforward Device."""
+
+    class DEVICES:
+        """."""
+
+        APU22_06SB = 'SI-06SB:ID-APU22'
+        APU22_07SP = 'SI-07SP:ID-APU22'
+        APU22_08SB = 'SI-08SB:ID-APU22'
+        APU22_09SA = 'SI-09SA:ID-APU22'
+        APU58_11SP = 'SI-11SP:ID-APU58'
+        EPU50_10SB = 'SI-10SB:ID-EPU50'
+        WIG180_14SB = 'SI-14SB:ID-WIG180'
+        ALL = (
+            EPU50_10SB, WIG180_14SB,
+            APU22_06SB, APU22_07SP, APU22_08SB, APU22_09SA, APU58_11SP,
+            )
+
+    def __init__(self, devname):
+        """."""
+        devname = _SiriusPVName(devname)
+
+        # check if device exists
+        if devname not in IDFF.DEVICES.ALL:
+            raise NotImplementedError(devname)
+
+        self._devname = devname  # needed for _create_devices
+        self._idffconfig = _IDFFConfig()
+
+        self._kparametername = \
+            _IDSearch.conv_idname_2_kparameter_propty(devname)
+
+        self._devkp, self._devsch, self._devscv, self._devsqs = \
+            self._create_devices(devname)
+
+        # call base class constructor
+        devices = [self._devkp, ]
+        devices += self._devsch
+        devices += self._devscv
+        devices += self._devsqs
+        super().__init__(devname=devname, devices=devices)
+
+    @property
+    def chnames(self):
+        """Return CH corrector power supply names."""
+        return _IDSearch.conv_idname_2_idff_chnames(self.devname)
+
+    @property
+    def cvnames(self):
+        """Return CV corrector power supply names."""
+        return _IDSearch.conv_idname_2_idff_cvnames(self.devname)
+
+    @property
+    def qsnames(self):
+        """Return QS corrector power supply names."""
+        return _IDSearch.conv_idname_2_idff_qsnames(self.devname)
+
+    @property
+    def kparametername(self):
+        """Return corresponding to ID kparameter."""
+        return self._kparametername
+
+    @property
+    def polarizations(self):
+        """Return list of possible light polarizations for the ID."""
+        return _IDSearch.conv_idname_2_polarizations(self.devname)
+
+    @property
+    def kparameter(self):
+        """Return kparameter value."""
+        return self._devkp[self._kparametername]
+
+    @property
+    def idffconfig(self):
+        """."""
+        return self._idffconfig
+
+    def load_config(self, name):
+        """Load IDFF configuration."""
+        self._idffconfig.load_config(name=name)
+        if not self._config_is_ok():
+            raise ValueError('Config is incompatible with ID configdb type.')
```

Comment:
the condition used here would be the configname? I think maybe this check should be done before loading the config...

Reply:
no, the idea here is to check whether the given configdb name for `si_idff` is compatible with the class instantiation for a specific ID name. the check needs to be performed after the config os loaded. we can not trust the config name.